src/main/java/br/com/tecnodev/controller/SubCategoryController.java

```diff
@@ -0,0 +1,99 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.server.ResponseStatusException;
+
+import javax.validation.Valid;
+import java.util.Comparator;
+import java.util.List;
+
+@Controller
+public class SubCategoryController {
+
+    private final SubCategoryRepository subCategoryRepository;
+    private final CategoryRepository categoryRepository;
+
+    public SubCategoryController(SubCategoryRepository subCategoryRepository, CategoryRepository categoryRepository) {
+        this.subCategoryRepository = subCategoryRepository;
+        this.categoryRepository = categoryRepository;
+    }
+
+    @GetMapping("/admin/subcategories/{code}")
+    public String listSubcategories(@PathVariable String code, Model model) {
+        Category category = categoryRepository.findByCode(code).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
+        List<SubCategory> subCategoryList = subCategoryRepository.getSubcategoryByCategoryCodeOrdered(code);
+        List<SubCategoryToListDTO> subcategoryDto = subCategoryList.stream().map(SubCategoryToListDTO::new).toList();
+        model.addAttribute("category", category);
+        model.addAttribute("subcategories", subcategoryDto);
+        return "subcategory/list";
+    }
+
+    @GetMapping("/admin/subcategories/new")
+    public String getSubcategoryForm(NewSubCategoryForm newSubCategoryForm, Model model) {
+        List<Category> categories = categoryRepository.findAll()
+                .stream()
+                .sorted(Comparator.comparing(Category::getName))
+                .toList();
+
+        model.addAttribute("category", categories);
+        model.addAttribute("subCategoryForm", newSubCategoryForm);
+        return "subcategory/insert";
+    }
+
+    @PostMapping("/admin/subcategories")
+    public String insertSubcategory(@Valid NewSubCategoryForm newSubCategoryForm, BindingResult result, Model model) {
+        Category category = categoryRepository.findById(newSubCategoryForm.getCategoryId()).orElseThrow(RuntimeException::new);
+        if (result.hasErrors()) {
+            return getSubcategoryForm(newSubCategoryForm, model);
+        }
+        subCategoryRepository.save(newSubCategoryForm.toEntity(category));
+        return listSubcategories(category.getCode(), model);
+    }
+
+    @GetMapping("/admin/subcategories/{catCode}/{subCode}")
+    public String getSubcategoryUpdateForm(@PathVariable String catCode, @PathVariable String subCode, NewSubCategoryFormUpdate subCategoryFormUpdate, Model model) {
+        SubCategory subCategory = subCategoryRepository.findByCode(subCode, catCode)
```

Comment:
Achei legal que você fez questão de procurar a subcategoria tanto por seu código quanto pelo código da categoria. Isso garante que não haja divergências de dados. No entanto, ainda acho válido esse método deixar claro isso. Acha que faz sentido?

```suggestion
        SubCategory subCategory = subCategoryRepository.findBySubcategoryCodeAndCategoryCode(subCode, catCode)
```